### PR TITLE
Initial documentation

### PR DESCRIPTION
**Pull request by nolar**
> Issue : #12 

Add some documentation to start with, including the concepts and Kopf's self-positioning in the ecosystem.

More pages will follow to this PR as soon as they are written.

This PR can be previewed at https://kopf.readthedocs.io/en/docs/

### CODE REVIEW

**Comment by nolar** [comment]
Finally, some consistent documentation with the examples and story-telling is ready.

There are also few additional pages & sections prepared locally — "as I would like to have it" style — but those ones require few extra functions in the code, and I prefer to do this after the tests are added.

File: docs/alternatives.rst [hunk 1]
```diff
@@ -0,0 +1,132 @@
+============
+Alternatives
+============
+
+Metacontroller
+==============
+
+The closest equivalent of Kopf is Metacontroller_.
```
Comment (samurang87) [inline docs/alternatives.rst]: "to" Kopf

File: docs/alternatives.rst [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+with no need for in-depth knowledge of Kubernetes or Go.
+
+However, it does that in a different way than Kopf does:
+with few yaml files describing the structure of your operator
+(beside the custom resource definition),
+and by wrapping your core domain logic into the Function-as-a-Service
+or into the in-cluster HTTP API deployments,
+which in turn react to the changes of the custom resources.
+
+As an operator developer, you still have to implement the infrastructure
```
Comment (samurang87) [inline docs/alternatives.rst]: it is usually recommended to avoid the "you" form, and use "one" instead :)

File: docs/alternatives.rst [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    which makes Kopf a *pythonic* framework in the first place,
+    not just *written with Python*.
+
+Kopf also makes the effort to keep the operator development human-friendly,
+which means at least the ease of debugging (e.g. with the breakpoints,
+running in a local IDE, not in the cloud), the logs readability,
+and other little pleasant things.
+
+And also Kopf allows to write *any* arbitrary domain logic of the resources,
+especially if it spans over the long time periods (hours, days if needed),
```
Comment (samurang87) [inline docs/alternatives.rst]: "spans over long time periods"

File: docs/alternatives.rst [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+and is not limited to the timeout restrictions of the HTTP APIs with their
+expectation of nearly-immediate outcome (i.e. in seconds or milliseconds).
+
+Metacontroller, however, is more mature, 1.5 years older than Kopf,
+and is backed by Google, which originally developed Kubernetes itself.
```
Comment (samurang87) [inline docs/alternatives.rst]: "Google, who"

File: docs/alternatives.rst [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+.. _k8s-operator: https://github.com/side8/k8s-operator
+
+
+CoreOS Operator SDK & Framework
+===============================
+
+`CoreOS Operator SDK`_ is not actually an operator framework.
+It is an SDK, i.e. a Software Development Kit,
+which generates the skeleton code for the operators-to-be,
+and then the developers should enrich it with the domain logic code as needed.
```
Comment (samurang87) [inline docs/alternatives.rst]: "and users should enrich" ... I guess here "developers" would mean the developers of the SDK :)

File: docs/alternatives.rst [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+`CoreOS Operator Framework`_, of which the abovementioned SDK is a part,
+is a bigger, more powerful, but very complicated tool for writing the operators.
```
Comment (samurang87) [inline docs/alternatives.rst]: "for writing operators"

File: docs/alternatives.rst [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+Both are developed purely for the Go-based operators.
```
Comment (samurang87) [inline docs/alternatives.rst]: "purely for Go-based"

File: docs/alternatives.rst [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+No other languages are supported.
+
+.. _CoreOS Operator SDK: https://github.com/operator-framework/operator-sdk
+.. _CoreOS Operator Framework: https://coreos.com/operators/
+
+From the CoreOS'es point of view, an operator is a method of packaging
+and managing a Kubernetes-native application (presumably of any purpose,
+such as MySQL, Postgres, Redis, ElasticSearch, etc) with Kubernetes APIs
+(e.g. the custom resources of ConfigMaps) and ``kubectl`` tooling.
+They refer to the operators as
```
Comment (samurang87) [inline docs/alternatives.rst]: "They refer to operators"

File: docs/alternatives.rst [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    * https://coreos.com/operators
+    * https://coreos.com/blog/introducing-operator-framework
+    * https://enterprisersproject.com/article/2019/2/kubernetes-operators-plain-english
+
+
+Helm Operator Kit
+=================
+
+`Helm App Operator Kit`_ is by its name not an operator framework, but an operator
+for the specific domain: deploying the Helm charts (packages).
```
Comment (samurang87) [inline docs/alternatives.rst]: "operating Helm charts"

File: docs/alternatives.rst [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+The arbitrary domain logic cannot be added (or it is not obvious how).
```
Comment (samurang87) [inline docs/alternatives.rst]: "An arbitrary domain"

File: docs/alternatives.rst [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+It is intended only for deplying the Helm-packaged application.
```
Comment (samurang87) [inline docs/alternatives.rst]: "deploying"

File: docs/async.rst [hunk 12]
```diff
@@ -0,0 +1,38 @@
+===========
+Async/Await
+===========
+
+.. todo:: Fit this page into the walk-through sample story?
+
+Kopf supports the asynchronous handler functions::
```
Comment (samurang87) [inline docs/async.rst]: "Kopf supports asynchronous..."

File: docs/async.rst [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+    import asyncio
+    import kopf
+
+    @kopf.on.create('zalando.org', 'v1', 'kopfexamples')
+    async def create_fn(spec, **_):
+        await asyncio.sleep(1.0)
+
+The async functions are a bit better than the regular functions,
```
Comment (samurang87) [inline docs/async.rst]: I'd rephrase this, to avoid making opinionated statements. E.g. "Async functions have the additional benefit over non-async ones to make the full stack trace available when exceptions occur or IDE breakpoints are used. "

File: docs/async.rst [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+since they are executed directly inside of the Kopf's own event loop
+in the main thread, which makes the full stack trace available
+for the exception stack traces and IDE breakpoints.
+
+The regular synchronous handlers, despite supported for convenience,
+are executed in the parallel threads (via the default executor of the loop),
```
Comment (samurang87) [inline docs/async.rst]: "executed in parallel threads"

File: docs/async.rst [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+    import asyncio
+    import kopf
+
+    @kopf.on.create('zalando.org', 'v1', 'kopfexamples')
+    async def create_fn(spec, **_):
+        await asyncio.sleep(1.0)
+
+The async functions are a bit better than the regular functions,
+since they are executed directly inside of the Kopf's own event loop
```
Comment (samurang87) [inline docs/async.rst]: "inside of Kopf's own event loop"

File: docs/async.rst [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+in the main thread, which makes the full stack trace available
+for the exception stack traces and IDE breakpoints.
+
+The regular synchronous handlers, despite supported for convenience,
```
Comment (samurang87) [inline docs/async.rst]: "Regular synchronous handlers, ..."

File: docs/cleanup.rst [hunk 17]
```diff
@@ -0,0 +1,11 @@
+=======
+Cleanup
+=======
+
+Once we have finished with the experiments, let's cleanup the cluster
```
Comment (samurang87) [inline docs/cleanup.rst]: A bit too informal :)

File: docs/concepts.rst [hunk 18]
```diff
@@ -0,0 +1,68 @@
+========
+Concepts
+========
+
+**Kubernetes** is a container orchestrator.
+Over time, Kubernetes has become a standard de facto for the enterprise
+infrastructure management, especially if it based on the microservices.
```
Comment (samurang87) [inline docs/concepts.rst]: "based on microservices"

File: docs/concepts.rst [hunk 19]
```diff
@@ -0,0 +1,68 @@
+========
+Concepts
+========
+
+**Kubernetes** is a container orchestrator.
+Over time, Kubernetes has become a standard de facto for the enterprise
```
Comment (samurang87) [inline docs/concepts.rst]: Opinionated statement :)

File: docs/concepts.rst [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+infrastructure management, especially if it based on the microservices.
+
+It provides some basic primitives to orchestrate the application deployments
```
Comment (samurang87) [inline docs/concepts.rst]: "to orchestrate application deployments"

File: docs/concepts.rst [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+on the low level ---such as the pods, jobs, deployments, services, ingresses,
```
Comment (samurang87) [inline docs/concepts.rst]: "on a low level"

File: docs/concepts.rst [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+persistent volumes and volume claims, secrets---
+and allows the Kubernetes cluster to be extended with the arbitrary
```
Comment (samurang87) [inline docs/concepts.rst]: "allows a Kubernetes cluster to be extended with arbitrary..."

File: docs/concepts.rst [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+custom resources and custom controllers.
+
+On the top level, it consists of Kubernetes API, through which the users
```
Comment (samurang87) [inline docs/concepts.rst]: "it consists of the Kubernetes API"

File: docs/concepts.rst [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+talk to Kubernetes, internal storage of the state of the objects (etcd),
+and a collection of the controllers. The command-line tooling (``kubectl``)
```
Comment (samurang87) [inline docs/concepts.rst]: "a collection of controllers"

File: docs/concepts.rst [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+can also be considered as a part of the solution.
+
+----
+
+**Kubernetes controller** is the logic (i.e. the behaviour) behind most
```
Comment (samurang87) [inline docs/concepts.rst]: "The Kubernetes controller"

File: docs/concepts.rst [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+of the objects, both built-in and added as the extension of Kubernetes.
```
Comment (samurang87) [inline docs/concepts.rst]: "behind most objects, both build-in and added as extension of Kubernetes"

File: docs/concepts.rst [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+For example, the ReplicaSet and Pod creation when a Deployment object
```
Comment (samurang87) [inline docs/concepts.rst]: "Examples of objects are ReplicaSet and Pod creation upon creation of a deployment, rolling version upgrades, ..."

File: docs/concepts.rst [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+is created, the rolling version upgrades, and so on.
+
+The main purpose of any controller is to bring the actual state
+of the cluster to the desired state, as expressed with the resources/object
+specifications.
+
+----
+
+**Kubernetes operator** is one kind of the controllers, which orchestrates
```
Comment (samurang87) [inline docs/concepts.rst]: "The kubernetes operation is one kind of controller..."

File: docs/concepts.rst [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+the objects of the specific kind, with some domain logic implemented in it.
```
Comment (samurang87) [inline docs/concepts.rst]: "orchestrates objects of a specific kind, with some domain logic implemented inside"

File: docs/concepts.rst [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+
+The essential difference between the operators and the controllers
```
Comment (samurang87) [inline docs/concepts.rst]: "between operators and controllers"

File: docs/concepts.rst [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+is that the operators are the domain-specific controllers,
+but not all controllers are necessary the operators:
+for example, the built-in controllers for pods, deployments, services, etc;
```
Comment (samurang87) [inline docs/concepts.rst]: it's unclear whether the built-in controllers are operators here :) (the sentence lacks a verb).

File: docs/concepts.rst [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+or the extensions of the object's life-cycles based on the labels/annotations.
+
+The essential similarity is that they both implement the same pattern:
+watching the objects and reacting to the objects' events (usually the changes).
+
+----
+
+**Kopf** is a framework to build the Kubernetes operators in Python.
```
Comment (samurang87) [inline docs/concepts.rst]: "to build Kubernetes operators"

File: docs/concepts.rst [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+
+As any framework, Kopf provides both the "outer" toolkit to run the operator,
+to talk to Kubernetes cluster, and to marshall the Kubernetes events
```
Comment (samurang87) [inline docs/concepts.rst]: "talk to the Kubernetes cluster"

Comment (samurang87) [reply docs/concepts.rst]: "marshal"

File: docs/concepts.rst [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+into the pure-Python functions of the Kopf-based operator,
+and the "inner" libraries to assist with a limited set of common tasks
+of manipulating the Kubernetes objects
+(however, it is not a yet another Kubernetes client library).
+
+.. seealso::
+
+    See :doc:`/architecture`
+    to understand how does Kopf work in details and what exactly does it do.
```
Comment (samurang87) [inline docs/concepts.rst]: "understand how Kopf works in detail, and what it does exactly"

File: docs/deployment.rst [hunk 35]
```diff
@@ -0,0 +1,131 @@
+==========
+Deployment
+==========
+
+Kopf can be executed out of the cluster, as long as the environment is
```
Comment (samurang87) [inline docs/deployment.rst]: "out of a cluster"

File: docs/deployment.rst [hunk 36]
```diff
@@ -0,0 +1,131 @@
...
+.. _RBAC: https://kubernetes.io/docs/reference/access-authn-authz/rbac/
+
+Here is an example of what a RBAC config should look like:
+
+.. literalinclude:: deployment-rbac.yaml
+    :caption: rbac.yaml
+    :name: rbac-yaml
+    :language: yaml
+
+And the  the created service account is attached to the pods as follows:
```
Comment (samurang87) [inline docs/deployment.rst]: typo, "the" twice

File: docs/errors.rst [hunk 37]
```diff
@@ -0,0 +1,92 @@
...
+    import kopf
+
+    @kopf.on.create('zalando.org', 'v1', 'kopfexamples')
+    def create_fn(spec, **_):
+        if not is_data_ready():
+            raise kopf.HandlerRetryError("The data is not yet ready.", delay=60)
+
+In that case, there is no need to sleep in the handler explicitly, thus blocking
+any other events, causes, and generally any other handlers on the same object
+from being handled (such as the deletion or the parallel handlers/sub-handlers).
```
Comment (samurang87) [inline docs/errors.rst]: "such as deletion or parallel handlers..."

File: docs/errors.rst [hunk 38]
```diff
@@ -0,0 +1,92 @@
...
+    as to be retried immediately, where it continues with the remaining
+    handlers.
+
+    The only difference is that this special case produces less logs.
+
+
+Fatal errors
+============
+
+If an exception raised inherits from `kopf.HandlerFatalError`, the handler
```
Comment (samurang87) [inline docs/errors.rst]: "if a raised exception"

File: docs/events.rst [hunk 39]
```diff
@@ -0,0 +1,93 @@
+======
+Events
+======
+
+.. warning::
+    Kubernetes itself contains a terminology conflict:
+    There are *events* when watching over the objects/resources,
+    such as in ``kubectl get pod --watch``.
+    And there are *events* as a built-on object kind,
```
Comment (samurang87) [inline docs/events.rst]: "is built-on" intentional?

File: docs/events.rst [hunk 40]
```diff
@@ -0,0 +1,93 @@
...
+    as shown in ``kubectl describe pod ...`` in the "Events" section.
+    In this documentation, they are distinguished as the watch-events
```
Comment (samurang87) [inline docs/events.rst]: distinguished as "watch-events" and "k8s-events".  This section describes k8s-events only"

File: docs/events.rst [hunk 41]
```diff
@@ -0,0 +1,93 @@
...
+    and the k8s-events. This section describes the k8s-events only.
+
+Handled objects
+===============
+
+.. todo:: body arg must be optional, meaning the currently handled object.
```
Comment (samurang87) [inline docs/events.rst]: "the body arg"

File: docs/events.rst [hunk 42]
```diff
@@ -0,0 +1,93 @@
...
+
+Kopf provides few tools to report the arbitrary information
```
Comment (samurang87) [inline docs/events.rst]: "provides some tools to report arbitrary information for the handled objects as Kubernetes events"

File: docs/events.rst [hunk 43]
```diff
@@ -0,0 +1,93 @@
...
+                   type='SomeType',
+                   reason='SomeReason',
+                   message='Some message')
+
+The type and reason are arbitrary, and can be anything.
+Some restrictions apply (e.g. no spaces).
+The message is also arbitrary free-text.
+However, newlines are not rendered nice and are not recommended.
+
+For convenience, few shortcuts are provided to mimic the Python's ``logging``::
```
Comment (samurang87) [inline docs/events.rst]: "a few shortcuts"

File: docs/events.rst [hunk 44]
```diff
@@ -0,0 +1,93 @@
...
+    def create_fn(body, **_):
+        kopf.event(body,
+                   type='SomeType',
+                   reason='SomeReason',
+                   message='Some message')
+
+The type and reason are arbitrary, and can be anything.
+Some restrictions apply (e.g. no spaces).
+The message is also arbitrary free-text.
+However, newlines are not rendered nice and are not recommended.
```
Comment (samurang87) [inline docs/events.rst]: "nicely" (but it's a bit vague, what happens?)

File: docs/events.rst [hunk 45]
```diff
@@ -0,0 +1,93 @@
...
+      Error     SomeReason  5s    kopf  Some exception: Exception text.
+      Warning   SomeReason  5s    kopf  Some message
+
+
+Other objects
+=============
+
+.. todo:: kubernetes and pykube objects should be accepted natively, not only the dicts.
+
+The events can be also attached to other objects, not only those handled
```
Comment (samurang87) [inline docs/events.rst]: "Events can also"

File: docs/events.rst [hunk 46]
```diff
@@ -0,0 +1,93 @@
...
+
+        pod = kubernetes.client.V1Pod()
+        api = kubernetes.client.CoreV1Api()
+        obj = api.create_namespaced_pod(namespace, pod)
+
+        msg = f"This pod is created by KopfExample {body['name']}"
+        kopf.info(obj.to_dict(), reason='SomeReason', message=msg)
+
+.. note::
+    The events are not persistent.
```
Comment (samurang87) [inline docs/events.rst]: "Events are not persistent"

File: docs/events.rst [hunk 47]
```diff
@@ -0,0 +1,93 @@
...
+    They are usually garbage-collected after some time, e.g. one hour.
+    All the reported information must be only for the short-term use.
```
Comment (samurang87) [inline docs/events.rst]: "for a short-term use"

File: docs/handlers.rst [hunk 48]
```diff
@@ -0,0 +1,231 @@
+========
+Handlers
+========
+
+.. todo:: Multiple handlers per script.
+
+The handlers are the Python functions with the actual behaviour
```
Comment (samurang87) [inline docs/handlers.rst]: "Handlers are Python functions"

File: docs/handlers.rst [hunk 49]
```diff
@@ -0,0 +1,231 @@
...
+of the custom resources.
+
+They are called when any custom resource (within the scope of the operator)
+is created, modified, or deleted.
+
+Any operator built with Kopf is based on the handlers.
```
Comment (samurang87) [inline docs/handlers.rst]: "is based on handlers"

File: docs/handlers.rst [hunk 50]
```diff
@@ -0,0 +1,231 @@
...
+
+To register a handler for an event, use the `kopf.on` decorator::
+
+    import kopf
+
+    @kopf.on.create('zalando.org', 'v1', 'kopfexamples')
+    def my_handler(spec, **_):
+        pass
+
+The variety of the available decorators is described below.
```
Comment (samurang87) [inline docs/handlers.rst]: "All available decorators are described below" :)

File: docs/handlers.rst [hunk 51]
```diff
@@ -0,0 +1,231 @@
...
+
+
+Arguments
+=========
+
+The following keyword arguments available to the handlers
```
Comment (samurang87) [inline docs/handlers.rst]: "arguments are available"

File: docs/handlers.rst [hunk 52]
```diff
@@ -0,0 +1,231 @@
...
+than the high-level causes::
+
+    import kopf
+
+    @kopf.on.event('zalando.org', 'v1', 'kopfexamples')
+    def something_happened(**kwargs):
+        pass
+
+
+In that case, the handlers progress is not tracked, and is not stored
```
Comment (samurang87) [inline docs/handlers.rst]: the handlers' or the handler's

File: docs/handlers.rst [hunk 53]
```diff
@@ -0,0 +1,231 @@
...
+
+    @kopf.on.event('', 'v1', 'pods')
+    def pod_changed(**kwargs):
+        pass
+
+
+Cause handlers
+==============
+
+Kopf goes further and beyond: it detects the actual causes of these events,
```
Comment (samurang87) [inline docs/handlers.rst]: lol

File: docs/handlers.rst [hunk 54]
```diff
@@ -0,0 +1,231 @@
...
+* Was the object just created?
+* Was the object deleted (marked for deletion)?
+* Was the object edited, and which fields specifically were edited,
+  from which old values to which new values?
+
+.. note::
+    Worth noting that Kopf stores the status of the handlers, such as their
+    progress or errors or retries, in the object itself (in its ``status``),
+    which triggers its own low-level events, but these events are not detected
+    as the separate causes, as there is nothing changed *essentially*.
```
Comment (samurang87) [inline docs/handlers.rst]: "as separate causes"

File: docs/handlers.rst [hunk 55]
```diff
@@ -0,0 +1,231 @@
...
+
+    @kopf.on.delete('zalando.org', 'v1', 'kopfexamples')
+    def my_handler(spec, **_):
+        pass
+
+
+Field handlers
+==============
+
+As a helper handler, the specific fields can be monitored
```
Comment (samurang87) [inline docs/handlers.rst]: "as helper handlers" I guess 🤔

File: docs/handlers.rst [hunk 56]
```diff
@@ -0,0 +1,231 @@
...
+There is no special detection of the causes for the fields,
+such as create/update/delete, so the field-handler is efficient
+only when the object is updated.
+
+
+Sub-handlers
+============
+
+.. warning::
+    Sub-handlers is an advanced topic. Please, make sure you understand
```
Comment (samurang87) [inline docs/handlers.rst]: "sub-handlers are an advanced topics"

File: docs/handlers.rst [hunk 57]
```diff
@@ -0,0 +1,231 @@
...
+    the regular handlers first, so as the handling cycle of the framework.
+
+As an extended concept, there are the sub-handlers.
```
Comment (samurang87) [inline docs/handlers.rst]: "Sub-handlers extend the concept of handlers" 🤔

File: docs/handlers.rst [hunk 58]
```diff
@@ -0,0 +1,231 @@
...
+
+A common example for this feature are the lists defined in the spec,
+each of which should be handled with a handler-like approach
+rather than explicitly -- i.e. with the error tracking, retries, logging,
+progress and status reporting, etc.
+
+This can be used with the dynamically created functions, such as the lambdas,
```
Comment (samurang87) [inline docs/handlers.rst]: "with dynamically created functions, such as lambdas"

File: docs/handlers.rst [hunk 59]
```diff
@@ -0,0 +1,231 @@
...
+partials (`functools.partial`), or the inner functions in the closures:
+
+.. code-block:: yaml
+
+    spec:
+      items:
+        - item1
+        - item2
+
+The sub-handlers can be implemented either imperatively::
```
Comment (samurang87) [inline docs/handlers.rst]: "Sub-handlers can be implemented"

File: docs/handlers.rst [hunk 60]
```diff
@@ -0,0 +1,231 @@
...
+    def create_fn(spec, **_):
+
+        for item in spec.get('items', []):
+
+            @kopf.on.this(id=item)
+            def handle_item(item=item, **_):
+
+                pass
+
+Both of these ways are equivalent. Actually, one is implemented via another.
```
Comment (samurang87) [inline docs/handlers.rst]: much bamboozle

File: docs/hierarchies.rst [hunk 61]
```diff
@@ -0,0 +1,83 @@
+===========
+Hierarchies
+===========
+
+One of the most common patterns of the operators is to create
+the children objects in the same Kubernetes cluster.
```
Comment (samurang87) [inline docs/hierarchies.rst]: "to create children objects"

File: docs/hierarchies.rst [hunk 62]
```diff
@@ -0,0 +1,83 @@
...
+    such as the official `kubernetes client`_ or pykube-ng_.
+
+.. _kubernetes client: https://github.com/kubernetes-client/python
+.. _pykube-ng: https://github.com/hjacobs/pykube
+
+
+Labelling
+=========
+
+To mark the created objects with the labels::
```
Comment (samurang87) [inline docs/hierarchies.rst]: "with labels"

File: docs/hierarchies.rst [hunk 63]
```diff
@@ -0,0 +1,83 @@
...
+
+This is commonly used for the parent objects that orchestrate multiple
+children objects of the same kind (e.g. as pods in the deployment).
+
+
+Same namespaces
+===============
+
+Usually, it is expected that the children objects are created in the same
+namespace, as their parent (unless there are strong reasons to do differently).
```
Comment (samurang87) [inline docs/hierarchies.rst]: "in the same namespace as their parent"

File: docs/loading.rst [hunk 64]
```diff
@@ -0,0 +1,27 @@
+Loading and importing
+=====================
+
+Kopf requires the source files with the handlers on the command line.
+It does not do any attempts to guess the user's intentions
+or to introduce any conventions (at least, now).
+
+There are two way to specify them (both mimicing the Python's own way):
```
Comment (samurang87) [inline docs/loading.rst]: mimicking

File: docs/loading.rst [hunk 65]
```diff
@@ -0,0 +1,27 @@
...
+    kopf run -m package1.module1 -m package2.module2
+
+* Or mixed::
+
+    kopf run file1.py file2.py -m package1.module1 -m package2.module2
+
+Which way to use depends on how the source code is structured,
+and is out of scope of Kopf.
+
+Each of the mentions files and modules will be imported.
```
Comment (samurang87) [inline docs/loading.rst]: "mentioned files and modules"

File: docs/naming.rst [hunk 66]
```diff
@@ -0,0 +1,18 @@
...
+
+Kopf is an abbreviation either for
+**K**\ubernetes **O**\perator **P**\ythonic **F**\ramework, or for
+**K**\ubernetes **OP**\erator **F**\ramework --- whatever you like more.
+
+"Kopf" also means "head" in German.
+
+It is capitalised in all natural language texts:
+
+    *I like to use Kopf to manage my domain in Kubernetes.*
```
Comment (samurang87) [inline docs/naming.rst]: "I like using Kopf"

File: docs/peering.rst [hunk 67]
```diff
@@ -0,0 +1,29 @@
+=======
+Peering
+=======
+
+All running operators communicate with each other via the peering objects
```
Comment (samurang87) [inline docs/peering.rst]: "via peering objects"

File: docs/peering.rst [hunk 68]
```diff
@@ -0,0 +1,29 @@
...
+(also the custom resources), so they know about each other.
```
Comment (samurang87) [inline docs/peering.rst]: "(also custom resources"

File: docs/peering.rst [hunk 69]
```diff
@@ -0,0 +1,29 @@
...
+
+The operator can be instructed to use the alternative peering objects::
```
Comment (samurang87) [inline docs/peering.rst]: "to use alternative peering objects"

File: docs/scopes.rst [hunk 70]
```diff
@@ -0,0 +1,19 @@
+======
+Scopes
+======
+
+By default, Kopf watches for the registered custom resources
```
Comment (samurang87) [inline docs/scopes.rst]: "for registered custom resources"

File: docs/scopes.rst [hunk 71]
```diff
@@ -0,0 +1,19 @@
...
+in the whole cluster. This may be not desired, for example when:
+
+* The cluster has restrictive permissions by the namespaces or by teams.
```
Comment (samurang87) [inline docs/scopes.rst]: "by namespaces"

File: docs/scopes.rst [hunk 72]
```diff
@@ -0,0 +1,19 @@
...
+* Different versions of the operator are deployed into different namespaces
+  to serve the custom resources only in those namespaces.
+
+
+Namespaces
+----------
+
+The running operator can be restricted to handle the custom resources
```
Comment (samurang87) [inline docs/scopes.rst]: "to handle custom resources"

File: docs/vision.rst [hunk 73]
```diff
@@ -0,0 +1,32 @@
+======
+Vision
+======
+
+Kubernetes is rising as a standard de facto for the infrastructure abstractions.
+
+Kubernetes operators become a common way to extend Kubernetes
+with the domain logic.
+
+However, at the moment (2018-2019), the operators are mostly written in Go,
```
Comment (samurang87) [inline docs/vision.rst]: "operators are mostly written"

File: docs/vision.rst [hunk 74]
```diff
@@ -0,0 +1,32 @@
+======
+Vision
+======
+
+Kubernetes is rising as a standard de facto for the infrastructure abstractions.
+
+Kubernetes operators become a common way to extend Kubernetes
```
Comment (samurang87) [inline docs/vision.rst]: "K8s operators have become"

File: docs/vision.rst [hunk 75]
```diff
@@ -0,0 +1,32 @@
...
+with the domain logic.
```
Comment (samurang87) [inline docs/vision.rst]: "with domain logic"

File: docs/vision.rst [hunk 76]
```diff
@@ -0,0 +1,32 @@
...
+
+However, at the moment (2018-2019), the operators are mostly written in Go,
+and require the advanced knowledge both of Go and of Kubernetes internals.
+This makes the Kubernetes operators a skill of few, a property of the "elites".
+
+In a perfect world of Kopf, the Kubernetes operators are a commodity,
```
Comment (samurang87) [inline docs/vision.rst]: "Kopf, Kubernetes operators"

File: docs/walkthrough/deletion.rst [hunk 77]
```diff
@@ -0,0 +1,73 @@
...
+
+        api = kubernetes.client.CoreV1Api()
+        obj = api.create_namespaced_persistent_volume_claim(
+            namespace=namespace,
+            body=data,
+        )
+
+        logger.info(f"PVC child is created: %s", obj)
+
+With this one line, `kopf.adopt` marks the PVC as the child of EVC.
```
Comment (samurang87) [inline docs/walkthrough/deletion.rst]: "marks the PVC as child of the EVC"

File: docs/walkthrough/diffs.rst [hunk 78]
```diff
@@ -0,0 +1,114 @@
+==================
+Diffing the fields
+==================
+
+Previously (:doc:`updates`), we have made the PVC size updated
```
Comment (samurang87) [inline docs/walkthrough/diffs.rst]: "we have set the size of the PVC to be updated every time..." ..." i.e. introduced a cascaded update"

File: docs/events.rst (hunk 39 above)
Comment (nolar) [reply docs/events.rst]: It is built-in.

File: docs/walkthrough/problem.rst [hunk 79]
```diff
@@ -0,0 +1,76 @@
...
+a little real-world problem with Kopf, step by step,
+presenting and explaining all the Kopf features one by one.
+
+
+Problem Statement
+=================
+
+In Kubernetes, there are no ephemeral volumes of big sizes, e.g. 500 GB.
+By ephemeral, it means that the volume does not persist after it is used.
+Such volumes can be used as the workspace for the large data crunching jobs.
```
Comment (samurang87) [inline docs/walkthrough/problem.rst]: "can be used as workspace for large data-crunching jobs"

File: docs/walkthrough/problem.rst [hunk 80]
```diff
@@ -0,0 +1,76 @@
...
+
+There is `Local Ephemeral Storage`__, which allocates some space on a node's
+root partition, shared with the docker images and other containers,
+but it is often limited in size depending on the node/cluster config:
```
Comment (samurang87) [inline docs/walkthrough/problem.rst]: mmm sounds familiar :D

File: docs/walkthrough/problem.rst [hunk 81]
```diff
@@ -0,0 +1,76 @@
...
+    spec:
+      containers:
+        - name: main
+          resources:
+            requests:
+              ephemeral-storage: 10G
+            limits:
+              ephemeral-storage: 10G
+
+There is `PersistentVolumeClaim`__, but they are persistent,
```
Comment (samurang87) [inline docs/walkthrough/problem.rst]: "There is a "persistentvolumeclaim" resource type, but it is persistent"

File: docs/walkthrough/problem.rst [hunk 82]
```diff
@@ -0,0 +1,76 @@
...
+================
+
+We will implement the ``EphemeralVolumeClaim`` object kind,
+which will be directly equivalent to ``PersistentVolumeClaim``
+(and will use it internally), but with a little extension:
+
+It will be *designated* for a pod or pods with a specific selection criteria.
+
+Once used, and all those pods are gone and are not going to be restarted,
+the ephemeral volume claim will be deleted after the *grace period*.
```
Comment (samurang87) [inline docs/walkthrough/problem.rst]: "after a grace period"

File: docs/walkthrough/resources.rst [hunk 83]
```diff
@@ -0,0 +1,91 @@
...
+Get a list of the existing objects of this kind with one of the commands:
+
+.. code-block:: bash
+
+    kubectl get EphemeralVolumeClaim
+    kubectl get ephemeralvolumeclaims
+    kubectl get evcs
+    kubectl get evc
+
+Please, note that we can use the short names as specified
```
Comment (samurang87) [inline docs/walkthrough/resources.rst]: the comma is not necessary

File: docs/walkthrough/updates.rst [hunk 84]
```diff
@@ -0,0 +1,91 @@
+====================
+Updating the objects
+====================
+
+Previously (:doc:`creation`),
+we have implemented a handler for the creation of a ``EphemeralVolumeClaim`` (EVC),
```
Comment (samurang87) [inline docs/walkthrough/updates.rst]: "an EphemeralVolumeClaim"

File: kopf/structs/progress.py [hunk 85]
```diff
@@ -35,14 +37,14 @@
...
+* ``status.kopf.failure`` are the handlers that failed completely (no retries).
+* ``status.kopf.delayed`` are the timestamps, until which these handlers sleep.
+* ``status.kopf.retries`` are number of retries for succeeded, failed,
   and for the progressing handlers.
 
 When the full event cycle is executed (possibly including multiple re-runs),
-the whole `status.kopf` section is purged. The life-long persistence of status
+the whole ``status.kopf`` section is purged. The life-long persistence of status
 is not intended: otherwise, multiple distinct causes will clutter the status
 and collide with the each other (especially critical for multiple updates).
```
Comment (samurang87) [inline kopf/structs/progress.py]: "with each other"

File: docs/handlers.rst [hunk 86]
```diff
@@ -0,0 +1,231 @@
...
+
+``**kwargs`` (or ``**_`` to stop the linting warnings on the unused variables)
+is required for the forward compatibility: the framework can add new keywords
+in the future, and the existing handlers should accept them and not break.
+
+
+Event handlers
+==============
+
+In some cases, it may be desired to watch the low-level raw events rather
```
Comment (samurang87) [inline docs/handlers.rst]: "watch low-level raw events rather than higher-level causes"

**Comment by nolar** [comment]
@samurang87 Thanks. All has been fixed (I hope). See the last commit (or a diff).
